### ISSUE TITLE
 ArgoCD generates irrelevant history records/sync results in monorepo setups during manual sync.

### ISSUE THREAD

**Issue by pasha-codefresh** [argoproj/argo-cd#17280]
When performing manual sync in a monorepo, it executes synchronization with the HEAD revision, which, in most cases, is not relevant to the application. 

To reproduce: 

Set up multiple apps like guestbook-1 and guestbook-2 in the same repo but different paths, all with the same version. Then, make a change only in guestbook-2 and manually sync guestbook-1. You'll notice it creates a new sync result and history record for no reason ( No actual diff for this path ) 

<img width="891" alt="Снимок экрана 2024-02-22 в 17 48 27" src="https://github.com/argoproj/argo-cd/assets/39732895/209789dd-e9cd-41e0-8f20-422fa0d0b6bf">
<img width="1485" alt="Снимок экрана 2024-02-22 в 17 48 32" src="https://github.com/argoproj/argo-cd/assets/39732895/d9a43714-b1bf-4b9d-b34d-71b9e64914ab">

You can see on screenshots that commit happened to guestbook2, but last sync result was updated for guestbook app, with autosync it would not happens because this check 

```
// Only perform auto-sync if we detect OutOfSync status. This is to prevent us from attempting
// a sync when application is already in a Synced or Unknown state
if syncStatus.Status != appv1.SyncStatusCodeOutOfSync {
    logCtx.Infof("Skipping auto-sync: application status is %s", syncStatus.Status)
    return nil, 0
}
```

To fix this, manual sync should work like autoSync. If there's no difference, it shouldn't do anything, preventing extra history records.

**Comment by agaudreault** [comment]
Seems like this would be fixed by https://github.com/argoproj/argo-cd/pull/15636

**Comment by pasha-codefresh** [comment]
Thank you @agaudreault , i will double check that it solves and issue and will start participate in PR review

**Comment by amotolani** [comment]
@agaudreault https://github.com/argoproj/argo-cd/pull/15636 does not actually fix this behaviour.  

AFAIK it only updates the cache with a new key if no changes are detected for the paths in the `argocd.argoproj.io/manifest-generate-paths` annotation, which is more for performance reasons. 

For manual sync, regardless of the status of the app (out of sync or not), a sync operation happens, adding to the history records.

**Comment by pasha-codefresh** [comment]
@amotolani yes, unfortunately it is expected behavior. I am working on dedicated controller that suppose to cover all rest cases , but it probably will be lab project and not part of argocd

**Comment by andrii-korotkov-verkada** [comment]
Is this still the case for the most recent argocd?

**Comment by seanorama** [comment]
> Is this still the case for the most recent argocd?

Yes. True for manual syncs. But also for automatic syncs if additional commits are made after the relevant commit but before the next ArgoCD polling.

And this is not a monorepo issue. It is more evident in large repos, but it can happen in the smallest of repos. If you change any file in repo, it will cause this inaccurate history. Such as a README.

**Comment by reegnz** [comment]
I've opened a PR for this issue. Most definitely looks like a bug to me.

**Comment by reegnz** [comment]
So over the past 2 weeks I was doing further analysis of the problem and here are my findings:

1. each app should have an option to filter commits to only specific paths, the ux is quite unusable the more unrelated commits (unrelated to the individual app) happen in the repo. The bigger the repo, the worse the signal-to-noise ratio in the sync history, but this is not strictly a monorepo-problem.
2. the operation we'd need to rely on to fix the history for each individual app is something like `git rev-list -1 -- {path}`. I don't know yet if it's possible to get a common set of commits if it's multiple paths involved with an app.
3. the git rev-list can be really slow with big repos. Based on my testing, [git commit-graph --changed-paths](https://git-scm.com/docs/git-commit-graph) solves for this performance issue quite well with bloom filters though.

So basically how I see it currently is that reposerver needs to gain functionality to generate a commit-graph with a changed-paths bloom filters and keep it updated whenever new changes roll in. Calculating is a bit expensive on first run, but incremental re-calculation is snappy.

Once reposerver has gained functionality to tell in a performant way what the last commit for a set of paths was, the functionality of providing meaningful hashes for the sync can then be built upon that reposerver functionality.
